### DISCUSSION TITLE
Who’s ready for Bubble Tea v2 Alpha?

### DISCUSSION THREAD

**Discussion by bashbunni** [charmbracelet/bubbletea#1156] [Announcements]
Hey everyone, we’re officially launching the Bubble Tea [`v2` alpha](https://github.com/charmbracelet/bubbletea/tree/v2.0.0-alpha.1). 

If you’re curious to know what the changes are and how to get started using it, check out the [release notes](https://github.com/charmbracelet/bubbletea/releases/tag/v2.0.0-alpha.1). The [examples](https://github.com/charmbracelet/bubbletea/tree/v2.0.0-alpha.1/examples) on that branch are also up-to-date with the new `v2` API changes, so you can _totally_ use those to guide you.

  What we would love from you is to have you **try it and give us feedback**. The best place to put your thoughts is here, in the announcement on GitHub. You can totally chat on Discord about it too, but the team can only really keep track of what’s on GitHub.   If you run into any issues, please file an [issue](https://github.com/charmbracelet/bubbletea/issues) on the Bubble Tea repo with `(v2)` in the title.  
  
  We’re super excited to share this with you all!

**Comment by dogwaterdev1** [comment]
YOOO LETS GOOOOO

**Comment by NachoNievaG** [comment]
Oh, this is going to be a wild ride 🚀

**Comment by blacktop** [comment]
I know there was a separate repo that added mouse support, but does v2 add native mouse support to bubbletea?

**Reply by bashbunni** [reply]
Hey @blacktop yes, Bubble Tea should have full mouse support in `v2` alpha. This is from the release notes
> We've also improved the mouse API, which is a breaking change. Use
tea.MouseMsg to match against different types of mouse events. Mouse messages
are split into tea.MouseClickMsg, tea.MouseReleaseMsg, tea.MouseWheelMsg,
and tea.MouseMotionMsg.

@aymanbagabas let me know if I'm missing any details here :)

**Reply by aymanbagabas** [reply]
Bubble Tea v1 and v2 both support mouse events.

**Reply by sa-** [reply]
Honestly that's fire

**Comment by ignoxx** [comment]
great changes 🚀

**Comment by Broderick-Westrope** [comment]
Awesome to see this progress. Unfortunately I haven't been able to get the `KeyRelease` to work in [this minimal example](https://gist.github.com/Broderick-Westrope/220e091d84803026fc934f264e8c1f4f). I'm using WezTerm version "20240203-110809-5046fc22" on MacOS.

Would appreciate any help :)

**Reply by meowgorithm** [reply]
Hey! I _think_ this is a known, and somewhat contested, issue in Alacritty (but @aymanbagabas correct me if I’m wrong)

https://github.com/alacritty/alacritty/issues/7435

**Reply by Broderick-Westrope** [reply]
Im using WezTerm. Are you suggesting it may also be an issue in WezTerm?

**Reply by bashbunni** [reply]
Hey, just to confirm, do you have `enable_kitty_keyboard=true` in your wezterm config? I'm just looking through the issues in the wezterm repo related to kitty keyboard https://github.com/search?q=repo%3Awez%2Fwezterm+kitty+keyboard&type=issues

**Reply by meowgorithm** [reply]
Apologies, @Broderick-Westrope I totally misread. I can confirm that `config.enable_kitty_keyboard = true` solves this in Wezterm (thanks @bashbunni).

**Reply by aymanbagabas** [reply]
I can confirm enabling kitty keyboard in the config solves the issue. Also related https://github.com/charmbracelet/bubbletea/pull/1161

**Comment by AdditionAddict** [comment]
new to charm (and go) but new api changes make so much more sense esp. key messages, got very confused by v1 how to do switch statements. love the changes!

**Comment by GeertJohan** [comment]
Very nice improvements!

Is there space to adress https://github.com/charmbracelet/bubbletea/discussions/434 in v2?
A major version is a good moment to do this because it probably needs a breaking change.

**Reply by meowgorithm** [reply]
Hi! There are a few different things being discussed in #434. Which one are you referring to exactly?

**Reply by GeertJohan** [reply]
The pointer receiver and consistent usage for the Update method. I make a proposl in my comment https://github.com/charmbracelet/bubbletea/discussions/434#discussioncomment-9919394

**Reply by bashbunni** [reply]
Hey @GeertJohan I just took a look at your linked comment. It's mostly directed at `bubbles` which aren't views in and of themselves, therefore don't need to actually implement the `tea.Model` interface. They are mainly meant to be reusable UI components.

We are planning on updating `bubbles` to make this distinction more clear and that will likely include updating them to use pointers. In `v2` none of the bubbles will implement `tea.Model`. I hope that helps to clarify. Let me know if anything is still unclear here :)

**Reply by GeertJohan** [reply]
Hey @bashbunni! I wanted to actually propose to change tea.Model but I see that its not very clear in my comment in that discussion. I did put the new signature it in the example, but more explicitly: I think the `(tea.Model).Update` method signature should be changed to drop the Model return value, thereby forcing a single possible.implementation of Model, namely via a pointer receiver. As I point out in my comment, the returned Model is confusing and has led to a number of different approaches when implementing the Model interface.

**Reply by GeertJohan** [reply]
The interface would then become:
```
type Model interface {
        Init() Cmd

        Update(Msg) (Cmd)

        View() string
}
```

**Reply by meowgorithm** [reply]
Your point are _very_ apt, though I will say we're hesitant to stray too far from Bubble Tea's functional roots, understanding completely the concern that Bubble Tea isn't idiomatic Go in places and also that Go is a decidedly imperative language.

A few questions:

1. Would this change require models to be pointers themselves? In other words, always `&model{}` and never `model{}`.
2. There are some very effective techniques that can come from being able to switch models entirely when returning, which I don't think is possible without being able to return a model. Am I correct here? For example:

```go
func (m Model) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
    if someCondition {
        return someOtherModel{}.Init()
    }
    return m, nil
}
```

**Reply by GeertJohan** [reply]
Hey @meowgorithm!

1. Implementations of Model can be stored as values (`var m = myModel{}`), but when they are passed to functions that expect the Model interface they must be passed as a pointer. This is no different from the current situation and is a result of how interfaces work. Example: https://go.dev/play/p/bK0FwkIYudk

2. That won't be possible.  It's ofcourse nice that this is currently possible with a small amount of code, but it also feels like it is unexpected of a model to do that. I would personally prefer to be more explicit about such a thing; using a wrapping model that switches from the first to the second model. There could be communication from the first model to the wrapper which causes the wrapper to swap it out for the second model. This will be more code, but also more explicit. And thereby easier to read, understand and maintain.

**Reply by qualidafial** [reply]
@GeertJohan 1. Your example only needs to use a pointer because your `FooModel.Update` method is implemented with a `(f *FooModel)` receiver. If I change the receiver to `(f FooModel)` then I can call `runUpdate(fooModelValue)` without taking `fooModelValue`'s address. https://go.dev/play/p/X4_prtoj9Z4.

With that said, there is nothing stopping you from using pointer receivers for your `tea.Model` implementations today, as long as your `Init`, `Update`, and `View` all use the same receiver type, and you pass the struct pointer to `tea.NewProgram()`.

**Reply by GeertJohan** [reply]
Hi @qualidafial I understand that the receiver beeds to be a pointer, but I don't think that's what @meowgorithm was referring to.

Also understand the pointer receiver is currently possible, this is more about removing ambiguity and increasing consistency. See #434

**Reply by qualidafial** [reply]
Understood. I do think that the similarity of `tea.Model` to `bubbles` component APIs like `list.Model` or `textinput.Model`, coupled with the fact they don't actually implement `tea.Model` was very confusing when I was getting started with bubbletea.

Possibly we could preserve the ability to switch to a new model via a special `tea.Cmd` e.g. `tea.SetModel(newModel)`

**Comment by qualidafial** [comment]
What is the team's v2 release plan wrt bubbles, lipgloss, and so on? Are there active v2 branches being worked on, or are we trying to nail down bubbletea first?

**Reply by meowgorithm** [reply]
We're planning on releasing `v2`s of Bubble Tea, Bubbles, Lip Gloss, and Huh all around the same time. All branches are in active development and we'll most likely see a `v2` alpha for Lip Gloss next.

**Comment by eduardoagarcia** [comment]
I'm excited for this! What's the ETA for this release?

**Reply by aymanbagabas** [reply]
Hey @eduardoagarcia! No ETA yet, but you can use the alpha version and tell us if you find any bugs or concerns

`go get github.com/charmbracelet/bubbletea/v2@v2.0.0-alpha.1`

**Comment by bashbunni** [comment]
Update on this, `v2-alpha.2` is live! https://github.com/charmbracelet/bubbletea/discussions/1237
